packages/contractkit/src/wrappers/TransactionBuilder.ts

```diff
@@ -0,0 +1,103 @@
+import { keccak256 } from 'ethereumjs-util'
+import Contract from 'web3/eth/contract'
+
+import { Address, CeloContract } from '../base'
+import { ContractKit } from '../kit'
+import { CeloTransactionObject, NumberLike, toBuffer } from './BaseWrapper'
+import { encodedTransactions, Transaction } from './Governance'
+
+export interface JSONTransaction {
+  value: NumberLike
+  celoContractName: CeloContract
+  methodName: string
+  args: any[]
+}
+
+type Web3Method = Contract['methods'][keyof Contract['methods']]
+export class TransactionBuilder {
+  public transactions: Transaction[]
+  private kit: ContractKit
+
+  constructor(_kit: ContractKit, _transactions: Transaction[] = []) {
+    this.transactions = _transactions
+    this.kit = _kit
+  }
+
+  get hash(): Buffer {
+    const encoded = this.encoded
+    return keccak256(
+      this.kit.web3.eth.abi.encodeParameters(
+        ['uint256[]', 'address[]', 'bytes', 'uint256[]'],
+        [encoded.values, encoded.destinations, encoded.data, encoded.dataLengths]
+      )
+    ) as Buffer
+  }
+
+  get encoded() {
+    return encodedTransactions(this.transactions)
+  }
+
+  appendCeloTx(tx: CeloTransactionObject<any>) {
```

Comment:
add another argument for tx arguments, same as `.send()` has within CeloTransactionObject. The user should be able to specify `value` for example. And not the `from`

Reply:
thinking this a bit more.

We should be able to extract: `data` and `to` from the `CeloTransactionObject`. same applies to the web `TransactionObject`.
as an optional extra argument we still need the `value`.

You shouldn't use `defaultParams` directly. This are optional and not required.

I suggest we add an accessor in CeloTransactionObject for `to` and `encodedAbi`.

How to implement those?

abi it's easy, just call `tx.encodeAbi()`. 

for  `to` it's tricky, the internal `txo` doesn't expose it publicly, but i've looked at the code and you should be able to do `txo._parent.options.address`

This is the actual web3 code:
```js
Contract.prototype._processExecuteArguments = function _processExecuteArguments(args, defer) {
    var processedArgs = {};

    processedArgs.type = args.shift();

    // get the callback
    processedArgs.callback = this._parent._getCallback(args);

    // get block number to use for call
    if(processedArgs.type === 'call' && args[args.length - 1] !== true && (_.isString(args[args.length - 1]) || isFinite(args[args.length - 1])))
        processedArgs.defaultBlock = args.pop();

    // get the options
    processedArgs.options = (_.isObject(args[args.length - 1])) ? args.pop() : {};

    // get the generateRequest argument for batch requests
    processedArgs.generateRequest = (args[args.length - 1] === true)? args.pop() : false;

    processedArgs.options = this._parent._getOrSetDefaultOptions(processedArgs.options);
    processedArgs.options.data = this.encodeABI();

    // add contract address
    if(!this._deployData && !utils.isAddress(this._parent.options.address))
        throw new Error('This contract object doesn\'t have address set yet, please set an address first.');

    if(!this._deployData)
        processedArgs.options.to = this._parent.options.address;

    // return error, if no "data" is specified
    if(!processedArgs.options.data)
        return utils._fireError(new Error('Couldn\'t find a matching contract method, or the number of parameters is wrong.'), defer.eventEmitter, defer.reject, processedArgs.callback);

    return processedArgs;
};
```
https://github.com/ethereum/web3.js/blob/ba00dbab7cb0000a76fe67dbaf246e23ad974271/packages/web3-eth-contract/src/index.js#L720

Also, `to` might be overrided by the defaultParams... but i wonder if we should allow that in the CeloTxObject

Reply:
can't seem to access `_parent` on the txo,
I came up with a "pipelined" design, let me know what you think